Readme.md
NodeJs Dropbox API Client
===

__node-dropbox__ is a Client providing communication between NodeJS and the __Dropbox API__ version 0 (and soon to be version 1).

The implementation's client contains full API coverage as in Dropbox's web docs, mobile docs might include some extra functionality and URLs; not discarded as to implemented soon. __node-oauth__ is the middleware used to perform the OAuth 1.0 requests.

Fully tested against Dropbox API (http://api.dropbox.com & http://api-content.dropbox.com).

Running Tests
===
Two options:

`$node test/index.js [testfile]`
   
`$node/test dropbox/[testfile]`

__Notes:__ To run the test file __client.js__ 2 arguments are required: the base64-encoded tokens key and token secret.

Usage
===
```javascript
var Auth = require('./node-dropbox/lib/auth');
var Client = require('./node-dropbox');

Auth.loadConfig('myConfigFile.json', function(err, config) {
   var options = {[...]};
   var oa = Auth.getInstance(options);
   var dropbox = new Client(apiHost, contentHost, port, oa, accessToken, accessTokenSecret);

   dropbox.getFile([...,] callback);
});
```

Public methods
===
```javascript
* accountInfo(callback)

* getFile(root, path, callback)

* putFile(root, path, content, [ctype], [callback])

* metadata(root, path, fileLimit, hash, list, callback)

* thumbnail(root, path, size, format, callback)

* copy(root, fromPath, toPath, callback)

* createFolder(root, path, callback)

* delete(root, path, callback)

* move(root, fromPath, toPath, callback)
```

Node Compatibility
===
0.4.x

__Notes:__ Further testing of previous versions is on the to-do list.

Author
===
Luis Merino &lt;mail AT luismerino DOT name&gt;

Feedback
===
Always highly appreciated. Please, use Github issues :)

License
===
[FreeBSD License](License.md)

History.md
0.1.1 / 2011-11-24
===
* Overhauled 'auth' test and test runner. Expanded documentation. Added FreeBSD License.

0.1.0 / 2011-11-19
===
* Initial release

test/index.js
#!/usr/local/bin/node

/*!
 * Node Dropbox test runner
 *
 * Copyright(c) 2011 Luis merino <mail AT luismerino DOT name>
 * FreeBSD License
 */

try {
    process.chdir(__dirname);
} catch(e) {
    throw e;
}

if (process.argv.length < 3)
    return console.error('\n1 argument required. Please, pass the test\'s name as first argument.\n');
    
var args = process.argv.splice(2);
var testName = args.shift().replace(/\-/g, '').trim().toLowerCase();

try {
    var testcase = require('./dropbox/' + testName);
} catch(e) {
    console.error('\n\033[33mError: first argument does not match with any named test.');
    var dirs = require('fs').readdirSync(__dirname + '/dropbox').filter(function(f) {
        return /\.js$/.test(f);
    });
    console.log('\n\033[36m- Available tests:\n\n    \033[37m* ' + dirs.join('\n    * '));
    console.log('\nExample: \033[32mnode [test|index.js] ' + dirs.shift() + '\n\033[0m');
    return;
}

testcase.argv = args;
testcase.exec();


test/helpers.js
/*!
 * Node Dropbox test 'Helpers' module
 *
 * Copyright(c) 2011 Luis merino <mail AT luismerino DOT name>
 * FreeBSD License
 */

//var Zombie = require('zombie/lib');
var Auth = require('../lib/auth');
var Client = require('../lib/client');
//var Https = require('https');
var QS = require('qs');
//var Url = require('url');
var Path = require('path');
var Spawn = require('child_process').spawn;
var assert = require('assert');
//var failed = new Error('Login failed');

module.exports = {
    
    newAuthenticator: function(config) {
        var options = {
            requestTokenUrl: config.requestTokenUrl,
            accessTokenUrl: config.accessTokenUrl,
            consumerKey: config.consumerKey,
            consumerSecret: config.consumerSecret,
        };
        
        return new Auth.getInstance(options);
    },
    
    newClient: function(config, accessToken, accessTokenSecret) {
        var apiHost = config.server;
        var contentHost = config.contentServer;
        var port = config.port;
        var auth = this.newAuthenticator(config);
        
        return new Client(apiHost, contentHost, port, auth, accessToken, accessTokenSecret);
    },
    
    /**
     * Python's headless browser: uses module 'mechanize.Browser'.
     * @ref Dropbox Python implementation tests.
     */
    loginAndAuthorize: function(authorizationUrl, config, callback) {
        Path.exists(__dirname + '/token_authorize.py', function(exists) {
            if (!exists)
                return callback(new Error('\'token_authorize.py\' file could not be found.'));
            
            var credentials = {
                testing_user: config.testingUser,
                testing_password: config.testingPassword
            };
            var args = [authorizationUrl, JSON.stringify(credentials)];
            var child = Spawn('./token_authorize.py', args);
            
            child.stdout.on('data', function(stream) {
                console.log(stream.toString());
            });
            child.stderr.on('error', function(stream) {
                delete child;
                callback(new Error(stream + ''));
            });
            child.on('exit', function() {
                if (child) {
                    child.kill();
                    callback();
                }
            });
        });
    }
    
    // Zombie (not working so far, throws JS compiling errors).
    /*loginAndAuthorize: function(token, config, callback) {
        var url = config.authorizationUrl + '?oauth_token=' + token;
        var parsedUrl = Url.parse(config.authorizationUrl);
        
        console.log('AUTHORIZING', url);
        
        var browser = new Zombie.Browser({ debug: true });
        browser.runScripts = false;
        browser.on('error', function(err) {
            console.log(err.stack);
        });
        // Load the authorization url
        browser.visit(url, function(err, browser) {
            // Fill email, password and submit form
            console.log('FIRST PAGE', browser.text('title') + '', browser.location.href);
            browser.runScripts = true;
            
            browser.
            fill('login_email', config.testingUser).
            fill('login_password', config.testingPassword).
            pressButton('login_submit', function(err, browser) {
                // Form submitted, new page loaded.
                console.log('RESULT URL', browser.location.href, browser.statusCode);
                
                browser.pressButton('allow_access', function(err) {
                    if (err)
                        return callback(err);
                    
                    console.log('TOKEN AUTHENTICATION ALLOWED!', token);
                    callback();
                });
            });
        });
    }*/
    
    // Custom mode (unknown mechanism, @todo dropbox team must reveal what their request really needs, in order for this to work).
    /*loginAndAuthorize: function(token, config, callback) {
        var url = Url.parse(config.authorizationUrl);
        console.log('AUTHORIZING', config.authorizationUrl, token);
        ;
        var options = {
            host: url.host,
            path: '/login?cont=' + config.authorizationUrl + '?oauth_token=' + token,
            headers: {}
        };
        
        // Enter login page
        var request = Https.get(options, function(response) {
            if (response.statusCode !== 200)
                return callback(failed);
            
            var data = '';
            response.on('data', function(chunk) {
                data += chunk;
            });
            response.on('end', proceed);
            
            function proceed() {
                var tVar = data.match(/name\=\"t\" value\=\"(.+)\"/);
                if (!tVar)
                    return callback(failed);
                
                tVar = tVar[1];
                var body = {
                    't': tVar,
                    'login_email': config.testingUser,
                    'login_password': config.testingPassword,
                    'remember_me': '1',
                    'login_submit': 'Log in',
                    'cont': config.authorizationUrl + '?oauth_token=' + token
                };
                body = QS.stringify(body);
                
                var cookies = [];
                response.headers['set-cookie'].forEach(function(str) {
                    var m = str.match(/^(.+?=.+?);/);
                    if (m && m[1])
                        cookies.push(m[1]);
                });
                
                options.path = '/login';
                options.method = 'POST';
                options.headers['Content-Length'] = body.length;
                options.headers['Content-Type'] = 'application/x-www-form-urlencoded';
                options.headers['Cookie'] = cookies.join('; ');
                
                // Submit to login page
                request = Https.request(options, function(response) {
                    if (response.statusCode !== 302)
                        return callback(failed);
                    
                    body = {
                        't': tVar,
                        'allow_access': 'Allow',
                        'oauth_token': token 
                    };
                    body = QS.stringify(body);
                    
                    var cookies2 = [];
                    [].concat(cookies).concat(response.headers['set-cookie']).forEach(function(str) {
                        var m = str.match(/^(.+?=.+?);/);
                        if (m && m[1] && m[1] != 'oscar=; Domain=www.dropbox.com' && m[1] != 'grouch=; Domain=www.dropbox.com')
                            cookies2.push(m[1] + '; expires=Sun, 11-Oct-2012 18:56:04 GMT; Path=/; secure; httponly');
                    });
                    
                    options.path = url.pathname;
                    options.headers['Accept-Charset'] = 'ISO-8859-1,utf-8;q=0.7,*;q=0.3';
                    options.headers['Accept-Language'] = 'en-US,en;q=0.8';
                    options.headers['Accept-Encoding'] = 'gzip,deflate,sdch';
                    options.headers['Cache-Control'] = 'max-age=0';
                    options.headers['Content-Length'] = body.length;
                    options.headers['Cookie'] = cookies2.join(';') + '; puc=YzYzOWFjMSxkNDNiZjQ0; __utmz=145599457.1317976548.5.2.utmcsr=google|utmccn=(organic)|utmcmd=organic|utmctr=create%20dropbox%20app; __utma=145599457.1483670903099351600.1317898674.1318010545.1318014643.10; __utmc=145599457; __utmb=145599457.2.10.1318014643';
                    options.headers['Cache-Control'] = 'max-age=0';
                    options.headers['Connection'] = 'keep-alive';
                    options.headers['Host'] = 'www.dropbox.com';
                    options.headers['Origin'] = 'https://www.dropbox.com';
                    options.headers['User-Agent'] = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_7_1) AppleWebKit/535.2 (KHTML, like Gecko) Chrome/15.0.874.83 Safari/535.2';
                    options.headers['Referer'] = 'https://www.dropbox.com/0/oauth/authorize?oauth_token=' + token;
                    
                    // Allow token authorization
                    request = Https.request(options, function(response) {
                        console.log(response.statusCode, response.headers);
                        
                        if (response.statusCode !== 500)
                            return callback(null, token);
                    });
                    
                    request.on('error', callback);
                    request.write(body);
                    request.end();
                });
                
                request.on('error', callback);
                request.write(body);
                request.end();
            }
        });
        
        request.on('error', callback);
        request.end();
    }*/
};

test/dropbox/auth.js
/*!
 * Node Dropbox 'Auth' test module
 *
 * Copyright(c) 2011 Luis merino <mail AT luismerino DOT name>
 * FreeBSD License
 */

require.paths.unshift(__dirname + '/../../support');

var Path = require('path');
var AsyncTest = require('async/lib/test');
var Auth = require('../../lib/auth');
var Helpers = require('../helpers');
var assert = require('assert');

/**
 * Test suite
 */
var tests = {
    
    name: 'auth',
    
    setUp: function(next) {
        if (!this.config) {
            var self = this;
            var configFile = Path.resolve(__dirname  + '/../../config/testing.json');
            
            Auth.loadConfig(configFile, function(err, config) {
                self.config = config;
                next();
            });
        }
        else
            next();
    },
    
    'test load config file': function(next) {
        var self = this;
        var configFile = Path.resolve(__dirname  + '/../../config/testing.json');
        
        Auth.loadConfig(configFile, function(err, config) {
            assert.ifError(err);
            assert.ok(config.server);
            assert.ok(config.contentServer);
            assert.equal(config.port, 80);
            assert.ok(config.requestTokenUrl);
            assert.ok(config.accessTokenUrl);
            assert.ok(config.authorizationUrl);
            assert.ok(config.trustedAccessTokenUrl);
            assert.ok(config.root === 'dropbox' || config.root === 'sandbox');
            assert.ok(config.appName);
            assert.ok(config.consumerKey);
            assert.ok(config.consumerSecret);
            assert.ok(config.testingUser);
            assert.ok(config.testingPassword);
            self.config = config;
            next();
        });
    },
    
    'test obtain oauth request token': function(next) {
        this.tokenPair = null;
        var oa = Helpers.newAuthenticator(this.config);
        var self = this;
        
        oa.getOAuthRequestToken(function(err, oAuthToken, oAuthTokenSecret) {
            assert.ifError(err);
            assert.ok(/[a-z0-9]{15}/.test(oAuthToken));
            assert.ok(/[a-z0-9]{15}/.test(oAuthTokenSecret));
            
            self.tokenPair = [oAuthToken, oAuthTokenSecret];
            next();
        });
    },
    
    'test authorize oauth request token': function(next) {
        var self = this;
        
        if (!this.tokenPair)
            this['test obtain oauth request token'](proceed);
        else
            proceed();
        
        function proceed() {
            assert.ok(self.tokenPair, 'Missing token pair');
            assert.notEqual(self.tokenPair.length, 0);
            var url = self.config.authorizationUrl + '?oauth_token=' + self.tokenPair[0];
            Helpers.loginAndAuthorize(url, self.config, next);
        }
    },
    
    'test obtain oauth access token': function(next) {
        var self = this;
        
        if (!this.tokenPair)
            this['test authorize oauth request token'](proceed);
        else
            proceed();
        
        function proceed() {
            var oa = Helpers.newAuthenticator(self.config);
            oa.getOAuthAccessToken(self.tokenPair[0], self.tokenPair[1],
                function(err, oAuthAccessToken, oAuthAccessTokenSecret) {
                    assert.ifError(err);
                    assert.ok(/[a-z0-9]{15}/.test(oAuthAccessToken));
                    assert.ok(/[a-z0-9]{15}/.test(oAuthAccessTokenSecret));
                    console.log('access_token: ' + oAuthAccessToken);
                    console.log('access_token_secret: ' + oAuthAccessTokenSecret);
                    next();
                });
        }
    }
};

var testcase = module.exports = AsyncTest.testcase(tests, 'node-dropbox', 20000);

if (require.main === module) {
    module.exports.exec();
}

lib/auth.js
/*!
 * Node Dropbox 'Auth' module.
 *
 * Wrapper for OAuth module instances. Includes utility function for loading
 * JSON file containing configuration parameters.
 *
 * Copyright(c) 2011 Luis merino <mail AT luismerino DOT name>
 * FreeBSD License
 */

/**
 * Module dependencies.
 */

var OAuth = require('node-oauth').OAuth;

/**
 * Fixes for POST requests which facilitates sending a body as a string, e.g.
 * 'multipart' and add extra oauth parameters at the same time. It removes the
 * previous blocking logic limited to just one or the other behavior.
 */

OAuth.prototype._putOrPost = function(method, url, oauth_token, oauth_token_secret, post_body, extra_params, post_content_type, callback) {
    if (typeof post_content_type === 'function') {
        callback = post_content_type;
        post_content_type = null;
    }
    
    if (!(typeof post_body === 'string' || Buffer.isBuffer(post_body))) {
        post_content_type = 'application/x-www-form-urlencoded';
        extra_params = post_body;
        post_body = null;
    }
    
    return this._performSecureRequest(oauth_token, oauth_token_secret, method, url, extra_params, post_body, post_content_type, callback);
};

OAuth.prototype.post = function(url, oauth_token, oauth_token_secret, post_body, extra_params, post_content_type, callback) {
    return this._putOrPost("POST", url, oauth_token, oauth_token_secret, post_body, extra_params, post_content_type, callback);
};

/**
 * Returns a new oauth instance using the passed configuration options.
 *
 * @param {options} object.
 * @return {OAuth} object
 */

exports.getInstance = function(options) {
    var defaults = {
        version: '1.0',
        signatureMethod: 'HMAC-SHA1'
    };
    
    var oa = new OAuth(
        options.requestTokenUrl,
        options.accessTokenUrl,
        options.consumerKey,
        options.consumerSecret,
        defaults.version,
        null,
        defaults.signatureMethod
    );
    
    return oa;
};

/**
 * Reads a config file and parses it as an object.
 *
 * @param {options} object.
 * @return {OAuth} object
 */

exports.loadConfig = function(path, callback) {
    require('fs').readFile(path, 'utf8', function(err, config) {
        if (err)
            return callback(err);
        
        var config;
        try {
            config = JSON.parse(config);
        } catch(e) {
            config = e;
        } finally {
            callback(null, config);
        }
    });
};

lib/util.js
/**
 * Node Dropbox 'Util' module
 * Copyright(c) 2011 Luis merino <mail@luismerino.name>
 * FreeBSD License
 */

/**
 * Module dependencies.
 */

var Sys = require('sys');
var EventEmitter = require('events').EventEmitter;

/**
 * Multipart class with purpose of forming multipart
 * body and headers strings for POST request.
 *
 * Completely based on the CoffeeScript example found at 
 * <http://fuzzycom.eu/posts/57> by Vincent Hellot <vincenthellot.tel>.
 *
 * @author Luis Merino <mail@luismerino.name>
 */

var Multipart = exports.Multipart = function(boundary, useBuffers) {
    EventEmitter.call(this);
    
    this.boundaryLength = 0;
    this.randomBoundaryPartLength = 32;
    this.boundary = boundary || this.getBoundary();
    this.crlf = '\r\n';
    this.useBuffers = useBuffers;
};

Sys.inherits(Multipart, EventEmitter);

(function() {
    
    // algorithm borrowed from curl Curl_FormBoundary
    this.getBoundary = function() {
        var allowedChars = '0123456789abcdef';
        var boundary = '';
        var i = 0;
        var len = this.boundaryLength - this.randomBoundaryPartLength;
        
        for (i = 0; i < len; i++)
            boundary += '-';
        
        for (i = 0; i < this.randomBoundaryPartLength; i++)
            boundary += allowedChars[Math.floor(Math.random() * 15 + 1)];
        
        return boundary;
    };
    
    this.write = function(str) {
        this.emit('data', this.useBuffers ? new Buffer(str, 'binary') : str);
    };
    
    this.addParameter = function(name, value, ctype) {
        this.write('--' + this.boundary);
        this.write(this.crlf);
        this.write('Content-Disposition: form-data; name="file"; filename="' + name + '"');
        this.write(this.crlf);
        this.write('Content-Type: ' + (ctype || 'text/plain')/* + '; charset=utf8'*/);
        this.write(this.crlf);
        this.write(this.crlf);
        this.write(value);
        this.write(this.crlf);
        
        return this;
    };
    
    this.end = function() {
        this.write('--' + this.boundary);
        this.write('--');
        this.emit('end');
    };
    
}).call(Multipart.prototype);

/**
 * String substitution function. Uses {value} pattern.
 *
 * Example:
 *   exports._t('www.{domain}.com', {domain: 'dropbox'}); // <- returns www.dropbox.com
 *
 * @param {str} String.
 * @param {object} Object contains the replacement values.
 * @returns String string with replacements applied.
 *
 * @author Luis Merino <mail@luismerino.name>
 */
 
exports._t = function(str, object) {
    return String(str).replace(/\\?\{([^{}]+)\}/g, function(match, name) {
        if (match.charAt(0) == '\\') return match.slice(1);
            return (object[name] != null) ? object[name] : '';
    });
};

/**
 * Utility function to check if the passed mime type behaves with binary encoding.
 *
 * @param {type} String mime type string.
 * @returns Boolean.
 *
 * @author Luis Merino <mail@luismerino.name>
 */

exports.isBinaryType = (function() {
    var mimeTypes = [
        'image/',
        'application/',
        'video/',
        'audio/'
    ];
    var dflt = 'text/plain';
    
    return function(type) {
        return mimeTypes.some(function(mtype) {
            return (type || dflt).match('^' + mtype);
        });
    }
})();

/**
 * From 'node-bufferjs' Buffer.concat by <AJ ONeal> (and Contributors).
 * Returns a new Buffer with the contents of b1, [...b2, ...b3],
 * or with the contents of all buffers in the Array.
 *
 * @see https://github.com/coolaj86/node-bufferjs
 */
 
(function() {
    "use strict";

    function concat(bufs) {
        var buffer, length = 0, index = 0;

        if (!Array.isArray(bufs)) {
            bufs = Array.prototype.slice.call(arguments);
        }
        for (var i=0, l=bufs.length; i<l; i++) {
            buffer = bufs[i];
            if (!Buffer.isBuffer(buffer)) {
                buffer = bufs[i] = new Buffer(buffer);
            }
            length += buffer.length;
        }
        buffer = new Buffer(length);

        bufs.forEach(function (buf, i) {
            buf = bufs[i];
            buf.copy(buffer, index, 0, buf.length);
            index += buf.length;
            delete bufs[i];
        });

        return buffer;
    }
    
    Buffer.concat = concat;

})();

lib/client.js
/*!
 * Node Dropbox 'Client' module
 *
 * Copyright(c) 2011 Luis merino <mail AT luismerino DOT name>
 * FreeBSD License
 */
 
module.exports.VERSION = '0.1.0';

/**
 * Module dependencies.
 */
 
var Util = require('./util');
var Multipart = Util.Multipart;

/**
 * Dropbox API version.
 */

var API_VERSION = module.exports.API_VERSION = 0;

/**
 * Class Client.
 *
 * @param {apiHost} string API host's path.
 * @param {contentHost} string API content host's path.
 * @param {post} number Connection port.
 * @param {auth} object OAuth instance.
 * @param {accessToken} string Pre-authorized access token hash.
 * @param {accessTokenSecret} string Access token secret hash.
 *
 * @return {Client} object
 */

var Client = module.exports = function(apiHost, contentHost, port, auth, accessToken, accessTokenSecret) {
    this.apiHost = apiHost;
    this.contentHost = contentHost;
    this.port = port;
    this.auth = auth;
    this.tokens = [accessToken, accessTokenSecret];
    this._init();
};

(function() {
    
    /**
     * Public methods
     */
    
    // Dropbox accounts
    this.accountInfo = function accountInfo(callback) {
        this.get(routes.gen(arguments), callback);
    };
    
    // Files and metadata
    this.getFile = function files(root, path, callback) {
        this._assertRoot(root);
        this._getRequestBody(this.get(routes.gen(arguments, 2)), callback);
    };
    
    this.putFile = function files(root, path, content, ctype, callback) {
        this._assertRoot(root);
        if (!callback) {
            callback = ctype;
            ctype = null;
        }
        
        // Extract the file name from path, using the last node separated by '/'.
        var name;
        path = path.split('/');
        name = path.pop();
        arguments[1] = path.join('/');
        
        var url = routes.gen(arguments, 2);
        var binary = Util.isBinaryType(ctype);
        var body = [];
        
        // Start multipart constructor, pass binary option if mime requires it.
        this.multipart = new Multipart(null, binary);
        
        this.multipart.on('data', function(data) {
            body.push(data);
        });
        this.multipart.on('end', function() {
            this.post(url, Buffer.concat(body), {file: name}, callback);
        }.bind(this));
        
        this.multipart.addParameter(name, content, ctype).end();
    };
    
    this.metadata = function metadata(root, path, fileLimit, hash, list, callback) {
        this._assertRoot(root);
        this.get(routes.gen(arguments, 5), callback);
    };
    
    this.thumbnail = function thumbnail(root, path, size, format, callback) {
        this._assertRoot(root);
        this._getRequestBody(this.get(routes.gen(arguments, 4)), callback);
    };
    
    // File operations
    this.copy = function copy(root, fromPath, toPath, callback) {
        this._assertRoot(root);
        this.post(routes.gen(arguments), {from_path: fromPath, to_path: toPath, root: root}, callback);
    };
    
    this.createFolder = function createFolder(root, path, callback) {
        this._assertRoot(root);
        this.post(routes.gen(arguments), {root: root, path: path}, callback);
    };
    
    this['delete'] = function _delete(root, path, callback) {
        this._assertRoot(root);
        this.post(routes.gen(arguments), {root: root, path: path}, callback);
    };
    
    this.move = function move(root, fromPath, toPath, callback) {
        this._assertRoot(root);
        this.post(routes.gen(arguments), {from_path: fromPath, to_path: toPath, root: root}, callback);
    };
    
    /**
     * Internal methods
     */
    
    this.get = function(url, callback) {
        return this.request('get', url, callback);
    };
    
    this.post = function(url, params, extraParams, callback) {
        if (typeof extraParams === 'function')
            callback = extraParams;
        else if (extraParams && typeof extraParams === 'object')
            params = {_params: params, _extra: extraParams};
        
        return this.request('post', url, params, callback);
    };
    
    this.request = function(method, url, params, callback) {
        if (['get', 'post'].indexOf(method) < 0)
            throw new Error('\'' + method.toUpperCase() + '\' requests are not allowed.');
        
        if (method === 'get') {
            callback = params;
            params = null;
        }
        
        var handler = callback
            ? (function(err, data, response) {
                return this._responseHandler(err, data, response, callback);
            }).bind(this)
            : null;
        
        if (method === 'get')
            return this.auth.get(this._buildUrl(url), this.tokens[0], this.tokens[1], handler);
        
        var encoding = null;
        if (params._params && params._extra) {
            encoding = 'multipart/form-data; boundary=' + this.multipart.boundary;
            return this.auth.post(this._buildUrl(url), this.tokens[0], this.tokens[1], params._params, params._extra, encoding, handler);
        }
        else
            return this.auth.post(this._buildUrl(url), this.tokens[0], this.tokens[1], params, null, handler);
    };
    
    /**
     * Private methods
     */
    
    this._responseHandler = function(err, data, response, callback) {
        var obj = {};
        try {
            obj = JSON.parse(data);
        }
        catch(e) {
            obj = null;
        }
        
        if (obj.error) {
            var msg = obj.error;
            var status;
            
            if (typeof msg === 'object')
                msg = JSON.stringify(msg);
            if (err.statusCode)
                status = err.statusCode;
            else {
                var match;
                if (match = msg.match(/^\s*(\d+)?\b/))
                    status = match[1];
            }
            if (!err instanceof Error)
                err = new Error();
            
            err.statusCode = status;
            err.message = msg || (data && data.message);
            
            callback(err, obj || data, response);
        }
        else
            callback(null, obj || data, response);
    };
    
    this._getRequestBody = function(request, callback) {
        request.on('response', function(response) {
            var data = [];
            //response.setEncoding('utf8');
            response.on('data', function(chunk) {
                data.push(chunk);
            });
            response.on('end', function() {
                afterRequest(response, data);
            });
        });
        request.on('error', function(err) {
            callback(err);
        });
        request.end();
        
        function afterRequest(response, data) {
            var binary = Util.isBinaryType(response.headers['content-type']);
            data = Buffer.concat(data);
            
            if (response.statusCode >= 200 && response.statusCode <= 299)
                callback(null, !binary ? data.toString('utf8') : data, response);
            else
                callback(new Error('Dropbox unknown error.'), data, response);
        }
    };
    
    this._buildUrl = function(path) {
        var fullUrl = path;
        
        if (!/^http:\/\//.test(path))
            fullUrl = (this.port === 80 ? 'http://' : 'https://') + path;
        
        return fullUrl;
    };
    
    this._assertRoot = function(root) {
        if (['dropbox', 'sandbox'].indexOf(root) < 0)
            throw new Error('Passed argument \'root\' must be \'dropbox\' or \'sandbox\'.');
    };
    
    this._init = function() {
        routes = routes.call(this);
    };
    
    var routes = function() {
        var ver = '/' + API_VERSION;
        var URL_MAP = {
            accountInfo: this.apiHost + ver + '/account/info',
            files: this.contentHost + ver + '/files/{root}{path}',
            metadata: this.apiHost + ver + '/metadata/{root}{path}?file_limit={file_limit}&hash={hash}&list={list}',
            thumbnail: this.contentHost + ver + '/thumbnails/{root}{path}?size={size}&format={format}',
            copy: this.apiHost + ver + '/fileops/copy',
            createFolder: this.apiHost + ver + '/fileops/create_folder',
            _delete: this.apiHost + ver + '/fileops/delete',
            move: this.apiHost + ver + '/fileops/move'
        };
        var _slice = Array.prototype.slice, x, i;
        
        for (x in URL_MAP) {
            i = 0;
            URL_MAP[x] = URL_MAP[x].replace(/\{[^{}]*\}/g, function() {
                return '{' + i++ + '}';
            });
        }
        
        URL_MAP.gen = function(args, n) {
            var fn = args.callee.name;
            if (undefined === n || null === n)
                return this[fn];
            
            var replacements = {};
            _slice.call(args, 0, n).forEach(function(arg, i) {
                replacements[i] = (undefined !== arg && null !== arg) ? arg : '';
            });
            return Util._t(this[fn], replacements);
        };
        
        return URL_MAP;
    };
        
}).call(Client.prototype);